File: java/src/hackerrank/ChocolateFeast.java
package hackerrank;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

// https://www.hackerrank.com/challenges/chocolate-feast/problem
public class ChocolateFeast {

    public static int chocolateFeast(int n, int c, int m) {
        if (n < c) return 0;
        int totalBars = n / c;
        int wrappersToExchange = totalBars;
        while (wrappersToExchange >= m) {
            int remainderWrapper = wrappersToExchange % m;
            wrappersToExchange /= m;
            totalBars += wrappersToExchange;
            wrappersToExchange += remainderWrapper;
        }
        return totalBars;
    }


    @Test
    @DisplayName("Test sample 0")
    void chocolateFeast0() {
        int[] n = { 10, 12, 6 };
        int[] c = { 2, 4, 2 };
        int[] m = { 5, 4, 2 };
        int[] expected = { 6, 3, 5 };
        for (int i =0; i < expected.length; i++) {
            int result = chocolateFeast(n[i], c[i], m[i]);
            assertEquals(expected[i], result);
        }
    }

}

File: kotlin/src/hackerrank/chocolateFeast.kt
package hackerrank

import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals


// https://www.hackerrank.com/challenges/chocolate-feast/problem

fun chocolateFeast(n: Int, c: Int, m: Int): Int {
    if (n < c) return 0
    var totalBars: Int = n / c
    var wrappersToExchange: Int = totalBars
    while (wrappersToExchange >= m) {
        val remainderWrapper: Int = wrappersToExchange % m
        wrappersToExchange /= m
        totalBars += wrappersToExchange
        wrappersToExchange += remainderWrapper
    }
    return totalBars
}


internal class ChocolateFeastTest {

    @Test
    @DisplayName("Test sample 0")
    fun chocolateFeast0() {
        val n: Array<Int> = arrayOf( 10, 12, 6 )
        val c: Array<Int> = arrayOf( 2, 4, 2 )
        val m: Array<Int> = arrayOf( 5, 4, 2 )
        val expected: Array<Int> = arrayOf( 6, 3, 5 )
        for (i in 0 until expected.size-1) {
            val result = chocolateFeast(n[i], c[i], m[i])
            assertEquals(expected[i], result)
        }
    }

}